backend/lib/spark-api-deal-submitter.js

```diff
@@ -0,0 +1,127 @@
+/** @import {PgPool, Queryable} from '@filecoin-station/deal-observer-db' */
+import Cursor from 'pg-cursor'
+
+/**
+ * Finds deals that haven't been submitted to the Spark API yet and submits them.
+ *
+ * @param {PgPool} pgPool
+ * @param {number} batchSize
+ * @param {(eligibleDeals: Array) => Promise<void>} submitDeals
+ */
+export const findAndSubmitUnsubmittedDeals = async (pgPool, batchSize, submitDeals) => {
+  console.debug(`Finding and submitting deals using batchSize: ${batchSize}`)
+  for await (const unsubmittedDeals of findUnsubmittedDeals(pgPool, batchSize)) {
+    console.debug(`Found ${unsubmittedDeals.length} unsubmitted deals`)
+    try {
+      const formattedDeals = formatUnsubmittedDealsForSparkApi(unsubmittedDeals)
+      await submitDeals(formattedDeals)
+      console.debug(`Successfully submitted ${formattedDeals.length} deals`)
+      await markDealsAsSubmitted(pgPool, unsubmittedDeals)
+    } catch (e) {
+      console.error('Failed to submit deals:', e)
+    }
+  }
+}
+
+/**
+ * Finds deals that haven't been submitted to spark api.
+ * For deals to be submitted they must
+ * - have not been submitted yet
+ * - were created more than 2 days ago
+ * - have payload cid
+ * - have not yet expired
+ *
+ * @param {PgPool} pgPool
+ * @param {number} batchSize
+ * @returns {AsyncGenerator<Array>}
+ */
+const findUnsubmittedDeals = async function * (pgPool, batchSize) {
+  const client = await pgPool.connect()
+  const cursor = client.query(new Cursor(`
+    WITH two_days_ago AS (
+        SELECT (NOW() - INTERVAL '2 days')::TIMESTAMP AS ts
+    )
+    SELECT
+        miner_id,
+        client_id,
+        piece_cid,
+        piece_size,
+        payload_cid,
+        epoch_to_timestamp(term_start_epoch + term_min) AS expires_at
+    FROM
+        active_deals
+    WHERE
+        submitted_at IS NULL
+        AND payload_cid IS NOT NULL
+        AND activated_at_epoch < timestamp_to_epoch((SELECT ts FROM two_days_ago))
+        AND epoch_to_timestamp(term_start_epoch + term_min) > NOW()`
+  ))
+
+  let rows = await cursor.read(batchSize)
+  while (rows.length > 0) {
+    yield rows
+
+    rows = await cursor.read(batchSize)
+  }
+
+  client.release()
+}
+
+/**
+ * Format unsubmitted deals to format expected by spark api.
+ *
+ * @param {Array} deals
+ * @returns {Array}
+*/
+const formatUnsubmittedDealsForSparkApi = (deals) => {
+  return deals.map(deal => ({
+    minerId: deal.miner_id,
+    clientId: deal.client_id,
+    pieceCid: deal.piece_cid,
+    pieceSize: deal.piece_size.toString(),
+    payloadCid: deal.payload_cid,
+    expiresAt: deal.expires_at
+  }))
+}
+
+/**
+ * Mark deals as submitted.
+ *
+ * @param {Queryable} pgPool
+ * @param {Array} eligibleDeals
+ */
+const markDealsAsSubmitted = async (pgPool, eligibleDeals) => {
+  await pgPool.query(`
+    UPDATE active_deals
+    SET submitted_at = NOW()
+    WHERE miner_id = ANY($1) AND client_id = ANY($2) AND piece_cid = ANY($3) AND piece_size = ANY($4)
```

Comment:
It's important to have an index in place to make this WHERE condition fast.

Please sync with @NikolasHaimerl about the UNIQUE index/constraint we want to add for active deals - see https://github.com/CheckerNetwork/deal-observer/pull/31/files#r1933931452

Depending on which columns uniquely identify a deal, you may need to add more of them to this WHERE condition.